libowfat/dns/dns_ip6.3
.TH dns_ip6 3
.SH NAME
dns_ip6 \- look up IPv6 addresses
.SH SYNTAX
.B #include <libowfat/dns.h>

int \fBdns_ip6\fP(stralloc* \fIout\fR,stralloc* \fIfqdn\fR);
.SH DESCRIPTION
dns_ip6 looks up 16-byte IPv6 addresses for the fully-qualified domain name in
\fIfqdn\fR. It puts the concatenation of the IPv6 addresses into \fIout\fR and
returns 0. If the domain does not exist in DNS, or has no IP addresses,
\fIout\fR will be empty.

dns_ip6 also looks up 4-byte IPv4 addresses and converts them into
IPv4-mapped IPv6 addresses (::ffff:1.2.3.4) to provide easy backwards
compatibility for IPv6 applications.

If dns_ip6 has trouble with the DNS lookup or runs out of memory, it returns
-1, setting errno appropriately. It may or may not change \fIout\fR.

If \fIfqdn\fR is an IPv6 address, dns_ip6 puts that IP address into
\fIout\fR without checking DNS.  IPv4 addresses are also used directly
without DNS lookup.
.SH "SEE ALSO"
dns_ip6_packet(3), dns_ip4(3), dns_name6(3)


libowfat/CHANGES
0.32:
  remove OpenBSD #warning (obsd maintainer says no longer needed)
  move headers to <libowfat/> upon install
  fix fmt_ip6 (Erwin Hoffmann)
  add MSG_ZEROCOPY support (only used for buffers >8k)
  use writev in buffer_put for a slight perf improvement
  add experimental iom API for multithreaded I/O multiplexing (in io.h)

0.31:
  special case buffer_get_token with token length 1 through memccpy (almost 4x speedup)
  test for pread and use it instead of lseek+read in cdb and io_mmapwritefile
  add mmap_readat (like mmap_read but open by fd, not by name)
  add fmt_iso8601 and scan_iso8601 (time_t -> "2014-05-27T19:22:16Z")
  add fmt_netstring and scan_netstring (http://cr.yp.to/proto/netstrings.txt)
  add (fmt|scan)_varint and (fmt|scan)_pb_* (binary marshaling ala protocol buffers)

0.30:
  add compiletimeassert.h
  add haveuint128.h
  byte.h now defines UNALIGNED_ACCESS_OK on x86
  add scan_ulongn, scan_8longn, scan_xlongn, scan_longn, scan_netstring
  change buffer to have a destructor function pointer
  SECURITY: fmt_strn would write one byte too many (returned right length though, so usually not a problem as that byte would be overwritten with \0 by the caller anyway)
  fmt_pad and fmt_fill fail more gracefully when srclen > maxlen
  You can now say $ make WERROR=-Werror (compiling t.c may fail but that's strictly speaking not part of the library)
  scan_html now decodes HTML entities based on the actual w3c list (from entities.json, say gmake update to fetch the current version)
  added fmt_escapechar* to fmt.h (implement various escaping mechanisms also found in textcode but for a single char not a whole string, and they always escape, not just when they think it's needed)
  scan_ushort was supposed to abort early and return 5 when attempting to parse "65536", because the result does not fit.  It did not.  Now it does.
  scan_*long, scan_*int, scan_*short now properly abort if the number would not fit
  SECURITY: check for integer overflow in stralloc_ready
  switch io_fds from array to newly implemented (hopefully thread-safe) iarray
  switch epoll from level triggering to edge triggering
  introduce io_eagain_read and io_eagain_write (discontinue using io_eagain plz)
  fix buffer_get
  add fmt_html_tagarg, fmt_xml, scan_html_tagarg
  add socket_fastopen, socket_fastopen_connect4, socket_fastopen_connect6 and socket_quickack
  on Linux, in iob_send, if MSG_MORE is defined, and the request type
    lends itself to it, used sendto/sendmsg with MSG_MORE instead of
    TCP_CORK (saves two syscalls)
  add io_fd_flags so the caller can tell io_fd whether the socket is blocking
    (saves one fcntl syscall)
  more constness for stralloc and buffer
  mmap_read/mmap_shared on zero length files no longer fail but return a
    zero length buffer
  if SOCK_NONBLOCK is defined, use it instead of socket+fcntl
  ... but if errno==EINVAL still fall back to socket+fcntl (Robert Henney)
  SECURITY: fix botched integer overflow handling logic in stralloc_ready (Giorgio)
  add critbit

0.29:
  save 8 bytes in taia.h for 64-bit systems
  add buffer_tosa (buffer writing to auto-growing stralloc)
  add iarray
  support SCTP in addition to TCP
  fix io_receivefd so the incoming buffer is only 1 byte; io_passfd
    sends only one byte, so we might receive (and discard) other data if
    we try to read more (Stefan Bühler, from the lightttpd project)
  add io_fd_canwrite (like io_fd but assume the fd is writable)
  save a few syscalls here and there
  add awesome hack in isset.h (comex, Linus Torvalds)
  add byte_equal_notimingattack

0.28:
  add uint64 pack and unpack routines
  fix subtle typo in sub_of (David Sirovsky)
  work around gcc bugs

0.27:
  add fmt_strm
  add iob_addbuf_munmap
  add socket_deferaccept
  catch one more case in umult64 (Tomi Jylhä-Ollila)

0.26:
  fix really pathological case where io_timeouted would never
    start over from the beginning because always new accept()ed
    connections came in and got newer, higher descriptors since the last
    io_timeouted loop.  (Dirk Engling)
  add some int overflow check macros to rangecheck.h
  fmt_ip6 compresses at best spot, not at first spot (Nikola Vladov)
  use inttypes.h to declare ints in uint*.h
  escape more in fmt_ldapescape
  try to catch malicious input in textcode fmt_* functions
  fmt_xlonglong was utterly broken (Johannes Vetter)

0.25:
  array_allocate no longer truncates the array
  array_get now not only checks whether the element fits in the
    allocates space, it also needs to be in the initialized space.
  add -D_REENTRANT to CFLAGS so libowfat can be used in multi-threaded
    programs
  further Windoze support (test/io5.c works, gatling still doesn't)
    This is just to get gatling to work, I may remove it again after
    that.
  implement Nikola's idea to remove limit number of strings in errmsg
  add taia_half
  add cdb
  add rangecheck.h
  add io_block
  make socket_(tc|ud)p[46] actually return non-blocking sockets as
    documented (Richard Lyons)
  fix for NetBSD 3.0 (Their /sys/types.h actually does not define u_char
    unless you #define _NETBSD_SOURCE, and then their CMSG_* macros use
    it.  This is horrible even by OpenBSD standards)
  remove support for ip6.int (it's no longer delegated)
  add asm versions of imult64 and umult64 for x86_64
    (22 cycles -> 12 cycles on my Athlon 64)
  use size_t and ssize_t instead of unsigned long et al
  add iob_bytesleft
  don't leak memory in test/httpd

0.24:
  fix scan_to_sa (Tim Lorenz)
  turns out we can not do file descriptor passing on OpenBSD
    if _XOPEN_SOURCE is defined, which is needed for Solaris.
    OpenBSD sucks.  I checked in a _really_ kludge now.  Please do not
    use OpenBSD if you have a choice.  Or Solaris, for that matter.
  turns out the imult routines (which I never used) were incorrect.
    Noted by Matthew Dempsky
  open_* from open.h now open in large file mode

0.23:
  also recognize EPFNOSUPPORT as EAFNOSUPPORT (groan)

libowfat-0.31/dns/dns_ip6.c
#include "stralloc.h"
#include "uint16.h"
#include "byte.h"
#include "dns.h"
#include "ip4.h"
#include "ip6.h"

static int dns_ip6_packet_add(stralloc *out,const char *buf,unsigned int len)
{
  unsigned int pos;
  char header[16];
  uint16 numanswers;
  uint16 datalen;

  pos = dns_packet_copy(buf,len,0,header,12); if (!pos) return -1;
  uint16_unpack_big(header + 6,&numanswers);
  pos = dns_packet_skipname(buf,len,pos); if (!pos) return -1;
  pos += 4;

  while (numanswers--) {
    pos = dns_packet_skipname(buf,len,pos); if (!pos) return -1;
    pos = dns_packet_copy(buf,len,pos,header,10); if (!pos) return -1;
    uint16_unpack_big(header + 8,&datalen);
    if (byte_equal(header,2,DNS_T_AAAA)) {
      if (byte_equal(header + 2,2,DNS_C_IN))
        if (datalen == 16) {
	  if (!dns_packet_copy(buf,len,pos,header,16)) return -1;
	  if (!stralloc_catb(out,header,16)) return -1;
	}
    } else if (byte_equal(header,2,DNS_T_A))
      if (byte_equal(header + 2,2,DNS_C_IN))
        if (datalen == 4) {
	  byte_copy(header,12,V4mappedprefix);
	  if (!dns_packet_copy(buf,len,pos,header+12,4)) return -1;
	  if (!stralloc_catb(out,header,16)) return -1;
	}
    pos += datalen;
  }

  dns_sortip6(out->s,out->len);
  return 0;
}

int dns_ip6_packet(stralloc *out,const char *buf,unsigned int len) {
  if (!stralloc_copys(out,"")) return -1;
  return dns_ip6_packet_add(out,buf,len);
}

static char *q = 0;

int dns_ip6(stralloc *out,stralloc *fqdn)
{
  unsigned int i;
  char code;
  char ch;
  char ip[16];

  if (!stralloc_copys(out,"")) return -1;
  if (!stralloc_readyplus(fqdn,1)) return -1;
  fqdn->s[fqdn->len]=0;
  if ((i=scan_ip6(fqdn->s,ip))) {
    if (fqdn->s[i]) return -1;
    stralloc_copyb(out,ip,16);
    return 0;
  }
  code = 0;
  for (i = 0;i <= fqdn->len;++i) {
    if (i < fqdn->len)
      ch = fqdn->s[i];
    else
      ch = '.';

    if ((ch == '[') || (ch == ']')) continue;
    if (ch == '.') {
      if (!stralloc_append(out,&code)) return -1;
      code = 0;
      continue;
    }
    if ((ch >= '0') && (ch <= '9')) {
      code *= 10;
      code += ch - '0';
      continue;
    }

    if (!dns_domain_fromdot(&q,fqdn->s,fqdn->len)) return -1;
    if (!stralloc_copys(out,"")) return -1;
    if (dns_resolve(q,DNS_T_AAAA) != -1)
      if (dns_ip6_packet_add(out,dns_resolve_tx.packet,dns_resolve_tx.packetlen) != -1) {
	dns_transmit_free(&dns_resolve_tx);
	dns_domain_free(&q);
      }
    if (!dns_domain_fromdot(&q,fqdn->s,fqdn->len)) return -1;
    if (dns_resolve(q,DNS_T_A) != -1)
      if (dns_ip6_packet_add(out,dns_resolve_tx.packet,dns_resolve_tx.packetlen) != -1) {
	dns_transmit_free(&dns_resolve_tx);
	dns_domain_free(&q);
      }
    return out->a>0?0:-1;
  }

  out->len &= ~3;
  return 0;
}
